modules/dbparser/tests/test_patterndb.c

```diff
@@ -706,434 +441,223 @@ test_simple_rule_with_rate_limited_action(void)
-      <actions>\
-        <action condition='\"$(context-length)\" eq \"1\"' trigger='match'>\
-          <message inherit-properties='TRUE'>\
-            <values>\
-              <value name='CONTEXT_LENGTH'>$(context-length)</value>\
-            </values>\
-          </message>\
-        </action>\
-      </actions>\
-    </rule>\
-    <rule provider='test' id='15' class='system' context-scope='program'\
-          context-id='$PID' context-timeout='60'>\
-      <patterns>\
-        <pattern>pattern15@ANYSTRING:p15@</pattern>\
-      </patterns>\
-      <actions>\
-        <action condition='\"$(context-length)\" eq \"2\"' trigger='match'>\
-          <message inherit-properties='FALSE'>\
-            <values>\
-              <value name='fired'>true</value>\
-            </values>\
-          </message>\
-        </action>\
-      </actions>\
-    </rule>\
-  </rules>\
- </ruleset>\
-</patterndb>";
+  assert_msg_matches_and_output_message_nvpair_equals(patterndb, "pattern13", 1, "CONTEXT_LENGTH", "2");
+  assert_msg_matches_and_output_message_nvpair_equals(patterndb, "pattern14", 1, "CONTEXT_LENGTH", "2");
 
-void
-test_patterndb_context_length(void)
-{
-  _load_pattern_db_from_string(pdb_msg_count_skeleton);
-
-  assert_msg_matches_and_output_message_nvpair_equals("pattern13", 1, "CONTEXT_LENGTH", "2");
-  assert_msg_matches_and_output_message_nvpair_equals("pattern14", 1, "CONTEXT_LENGTH", "2");
-
-  assert_msg_with_program_matches_and_nvpair_equals("prog2", "pattern15-a", "p15", "-a");
+  assert_msg_with_program_matches_and_nvpair_equals(patterndb, "prog2", "pattern15-a", "p15", "-a");
 
   _dont_reset_patterndb_state_for_the_next_call();
-  assert_msg_matches_and_output_message_nvpair_equals("pattern15-b", 2, "fired", "true");
+  assert_msg_matches_and_output_message_nvpair_equals(patterndb, "pattern15-b", 2, "fired", "true");
 
-  _destroy_pattern_db();
+  _destroy_pattern_db(patterndb);
 }
 
-gchar *tag_outside_of_rule_skeleton = "<patterndb version='3' pub_date='2010-02-22'>\
- <ruleset name='testset' id='1'>\
-  <patterns>\
-   <pattern>prog1</pattern>\
-  </patterns>\
-  <tags>\
-   <tag>tag1</tag>\
-  </tags>\
- </ruleset>\
-</patterndb>";
+//#include "test_parsers_e2e.c"
 
-void
-test_patterndb_tags_outside_of_rule(void)
+typedef struct _patterndb_test_param
 {
-  patterndb = pattern_db_new();
-  messages = NULL;
+  const gchar *pattern_db_file;
+  const gchar *program;
+  const gchar *message;
+  const gchar *name;
+  const gchar *expected_value;
+} PatternDBTestParam;
 
-  g_file_open_tmp("patterndbXXXXXX.xml", &filename, NULL);
-  g_file_set_contents(filename, tag_outside_of_rule_skeleton,
-                      strlen(tag_outside_of_rule_skeleton), NULL);
+ParameterizedTestParameters(pattern_db, test_rules)
+{
+  static PatternDBTestParam parser_params[] =
+  {
+    // test_conflicting_rules_with_different_parsers
+    {
+      .pattern_db_file = PATH_PREFIX"pdb_conflicting_rules_with_different_parsers.xml",
```

Comment:
I think this points toward a wrong direction. The unit test should not have external files that they are depending on. 
I know that the existing version created and loaded that file (but I think that was less evil.)

This is probably out of current PR's scope, just to give you a hint, where I think it should move and why is it important not to imply the current is the good direction.

The `PatternDB` itself should not know where its contents come: file, user input, stream or whatever. Try to think how you would modify the `PatternDB` if you would want to add a feature like embedded `PatternDB` in syslog-ng configuration (like `python`).

```
patterndb inlinepattern {
 <patterndb version='5' pub_date='2010-02-22'>
 ... 
</patterndb>
};
```

Actually this is exactly what we would like to achieve in the unit test IMHO.

(Anyway if this is kept as it is - :disappointed: , those files should be added to the `EXTRA_DIST` target.)

Reply:
I don't really catch the point. I need some more details about your concern. let's discuss it personally...

Reply:
I removed the XML files and moved their content into test_patterndb.h header file.